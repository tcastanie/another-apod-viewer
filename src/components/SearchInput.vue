<template>
  <form>
    <div class="grid">
      <label for="date1">
        {{
          isRange
            ? "Enter first date of the interval"
            : "Go to specific date picture"
        }}
        <input
          id="date1"
          v-model="searchInput1"
          name="date1"
          type="date"
          min="1995-06-16"
        />
      </label>
      <div v-if="!isRange"></div>
      <label v-if="isRange" for="date2">
        Enter second date of the interval
        <input
          id="date2"
          v-model="searchInput2"
          name="date2"
          type="date"
          min="1995-06-20"
        />
      </label>
    </div>
    <fieldset>
      <label for="range">
        <input
          id="range"
          v-model="isRange"
          type="checkbox"
          name="range"
          role="switch"
        />
        Search for a range of dates
      </label>
    </fieldset>
    <button
      :disabled="isRange && rangeBetweenDates > 20"
      @click.prevent="startSearch"
    >
      {{
        isRange
          ? rangeBetweenDates > 20
            ? "❌ Range is too big ! (> 20)"
            : "🔍 Search"
          : "🪐 Go"
      }}
    </button>
  </form>
  <LoadingLayer v-if="loading" />
  <ErrorLayer v-if="errorMessage">{{ errorMessage }}</ErrorLayer>
</template>

<script>
import { useStore } from "@/stores/apod";
import LoadingLayer from "./LoadingLayer.vue";
import ErrorLayer from "./ErrorLayer.vue";
export default {
  components: { LoadingLayer, ErrorLayer },
  emits: ["searchResults"],
  setup() {
    const store = useStore();
    return { store };
  },
  data() {
    return {
      isRange: false,
      searchInput1: "",
      searchInput2: "",
      errorMessage: null,
      loading: false,
    };
  },
  computed: {
    dateFourDaysBefore() {
      const date = new Date();
      const day = date.getDate();
      date.setDate(day - 4);
      return date.toISOString().split("T")[0];
    },
    dateYesterday() {
      const date = new Date();
      date.setDate(date.getDate() - 1);
      return date.toISOString().split("T")[0];
    },
    rangeBetweenDates() {
      const date1 = new Date(this.searchInput1);
      const date2 = new Date(this.searchInput2);
      const diff = Math.abs(date2.getTime() - date1.getTime());
      const days = Math.ceil(diff / (1000 * 3600 * 24));
      return days;
    },
  },
  mounted() {
    this.searchInput1 = this.dateFourDaysBefore;
    this.searchInput2 = this.dateYesterday;
  },
  methods: {
    startSearch() {
      if (!this.isRange) {
        this.$router.push({
          name: "date",
          params: { date: this.searchInput1 },
        });
      } else {
        this.loading = true;
        const url = this.store.getUrlWithRange(
          this.searchInput1,
          this.searchInput2
        );
        fetch(url)
          .then(async (response) => {
            const data = await response.json();
            if (!response.ok) {
              const error1 = data && (data.msg || data.error.code);
              const error2 = `${response.statusText} (${response.status})`;
              return Promise.reject(new Error(`${error1}. ${error2}`));
            }
            this.loading = false;
            this.$emit("searchResults", data);
          })
          .catch((error) => {
            this.errorMessage = error;
            this.loading = false;
          });
      }
    },
  },
};
</script>
