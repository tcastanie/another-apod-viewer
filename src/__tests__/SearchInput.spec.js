import { beforeEach, describe, expect, it } from "vitest";
import { createRouter, createWebHistory } from "vue-router";
import { flushPromises, mount } from "@vue/test-utils";
import SearchInput from "@/components/SearchInput.vue";
import { createTestingPinia } from "@pinia/testing";
import { routes } from "@/router/index.js";

let router;
beforeEach(async () => {
  router = createRouter({
    history: createWebHistory(),
    routes,
  });
});

const yesterday = new Date();
yesterday.setDate(yesterday.getDate() - 1);
const yesterdayDate = yesterday.toISOString().split("T")[0];
const fourDaysBefore = new Date();
fourDaysBefore.setDate(fourDaysBefore.getDate() - 4);
const fourDaysBeforeDate = fourDaysBefore.toISOString().split("T")[0];

describe("SearchInput", () => {
  it("renders properly", () => {
    const wrapper = mount(SearchInput, {
      global: {
        plugins: [createTestingPinia()],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("has correct values", async () => {
    const wrapper = mount(SearchInput, {
      global: {
        plugins: [createTestingPinia()],
      },
    });
    expect(wrapper.vm.searchInput1).toBe(fourDaysBeforeDate);
    expect(wrapper.vm.searchInput2).toBe(yesterdayDate);
    expect(wrapper.vm.rangeBetweenDates).toBe(3);
    expect(wrapper.find("button").attributes("disabled")).toBeUndefined();
    expect(wrapper.find("button").text()).toContain("Go");

    const date1 = "2020-01-01";
    const date2 = "2020-01-08";
    await wrapper.setData({
      searchInput1: date1,
      searchInput2: date2,
    });
    expect(wrapper.vm.rangeBetweenDates).toBe(7);

    await wrapper.find("#range").setValue(true);
    await wrapper.find("#date1").setValue("2022-01-01");
    await wrapper.find("#date2").setValue("2022-01-12");
    expect(wrapper.vm.searchInput1).toBe("2022-01-01");
    expect(wrapper.vm.searchInput2).toBe("2022-01-12");
    expect(wrapper.vm.rangeBetweenDates).toBe(11);
  });

  it("disable search button correctly", async () => {
    const wrapper = mount(SearchInput, {
      global: {
        plugins: [createTestingPinia()],
      },
      computed: {
        rangeBetweenDates() {
          return 25;
        },
      },
    });
    await wrapper.find("#range").setValue(true);
    expect(wrapper.find("button").attributes("disabled")).toBe("");
    expect(wrapper.find("button").text()).toContain("Range is too big");
  });

  it("open singnle apod correctly", async () => {
    router.push("/search");
    await router.isReady();
    const wrapper = mount(SearchInput, {
      global: {
        plugins: [router, createTestingPinia()],
      },
    });
    await wrapper.find("button").trigger("click");
    await flushPromises();
    expect(wrapper.vm.$route.path).toBe(`/${fourDaysBeforeDate}`);
  });

  it("search apods in range correctly", async () => {
    router.push("/search");
    await router.isReady();
    const wrapper = mount(SearchInput, {
      global: {
        plugins: [router, createTestingPinia({ stubActions: false })],
      },
    });
    await wrapper.find("#range").setValue(true);
    await wrapper.find("button").trigger("click");
    await flushPromises();
    // Check if the apods are correctly loaded
  });
});
