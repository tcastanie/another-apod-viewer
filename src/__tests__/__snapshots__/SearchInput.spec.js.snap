// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchInput > renders properly 1`] = `
"<form>
  <div class=\\"grid\\"><label for=\\"date1\\">Go to specific date picture <input id=\\"date1\\" name=\\"date1\\" type=\\"date\\" min=\\"1995-06-16\\"></label>
    <div></div>
    <!--v-if-->
  </div>
  <fieldset><label for=\\"range\\"><input id=\\"range\\" type=\\"checkbox\\" name=\\"range\\" role=\\"switch\\"> Search for a range of dates </label></fieldset><button>🪐 Go</button>
</form>
<!--v-if-->
<!--v-if-->"
`;
